### PR TITLE
[mplex] Add benchmark, tweak default `split_send_size`.

### PR DESCRIPTION
**Pull request by romanb**
This PR supersedes https://github.com/libp2p/rust-libp2p/pull/802. It includes  a benchmark that sends _and_ receives __1MiB__ of data via __TCP__ and __memory__ transports with varying choices for the `split_send_size`. Always using the plaintext protocol for "authentication". Typical results from my side look as follows:
```
tcp/256                 time:   [1.3490 ms 1.3550 ms 1.3619 ms]  
tcp/512                 time:   [1.0430 ms 1.0493 ms 1.0563 ms]                     
tcp/1024                time:   [788.28 us 792.05 us 795.92 us]                     
tcp/8192                time:   [609.95 us 614.46 us 619.14 us]                     
tcp/16384               time:   [586.30 us 592.73 us 600.07 us]                      
tcp/65536               time:   [571.13 us 575.46 us 580.18 us]                      
tcp/262144              time:   [587.14 us 593.29 us 599.73 us]                       
tcp/1048576             time:   [2.1731 ms 2.2119 ms 2.2529 ms]                         

memory/256              time:   [1.2023 ms 1.2125 ms 1.2231 ms]                        
memory/512              time:   [723.83 us 730.71 us 738.00 us]                       
memory/1024             time:   [466.72 us 471.63 us 477.29 us]                        
memory/8192             time:   [289.08 us 293.97 us 299.65 us]                        
memory/16384            time:   [270.09 us 274.39 us 278.91 us]                         
memory/65536            time:   [248.38 us 254.62 us 261.55 us]                         
memory/262144           time:   [273.51 us 281.88 us 290.79 us]    
```

Based on such numbers, and probably not surprisingly, __8KiB__ seems to be a good default, as it also is the buffer size of the underlying `Framed` codec. Hence this PR changes the default from `1KiB` to `8KiB`.

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
I am in favor of the data driven approach.

`split_send_size` limits the maximum amount of data a substream can send in one go. Do I understand correctly that the underlying motivation for this mechanism is to control the amount of latency a substream can induce on another substream, by enforcing interleaving at a certain packet size?

**Comment by romanb** [comment]
> Do I understand correctly that the underlying motivation for this mechanism is to control the amount of latency a substream can induce on another substream, by enforcing interleaving at a certain packet size?

As far as I'm concerned, the motivation is not primarily better interleaving of substream write operations, though that may be a side-effect. My reasoning here is two-fold:

  1. We want to scope a single substream write to writing of a single Mplex frame. A single frame cannot be arbitrarily large, the maximum is `1MiB` per the spec. We don't want to return errors when a write operation tries to write more than that but rather write as much as we can, so the given data to write needs to be cut somewhere.
  2. We could always cut at `1MiB`, i.e. trying to always produce maximally sized frames (to compare, I think yamux does so w.r.t. the remaining window size, which is up to `256KiB` by default). However, it does not seem to be beneficial to take 1MiB-sized "bites" from the input if it then immediately needs to be broken down into smaller chunks which themselves need to be potentially buffered again in intermediate layers, which I think is what happens in the underlying `Framed` (8KiB default buffer size that grows / allocates as necessary) and further down the TCP stack. I couldn't get too much out of the flamegraphs I shortly looked at, but I think this is at least part of the reason why the benchmark deteriorates again with larger Mplex frame sizes towards `1MiB`.

File: muxers/mplex/benches/split_send_size.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+];
+
+fn prepare(c: &mut Criterion) {
+    let _ = env_logger::try_init();
+
+    let payload: Vec<u8> = vec![1; 1024 * 1024 * 1];
+
+    let mut tcp = c.benchmark_group("tcp");
+    let tcp_addr = multiaddr![Ip4(std::net::Ipv4Addr::new(127,0,0,1)), Tcp(0u16)];
+    for &size in BENCH_SIZES.iter() {
```
[the remark this answers is not in the document]
Comment (romanb) [reply muxers/mplex/benches/split_send_size.rs:55]: I wasn't aware of that and will look into using it, thanks for the suggestion!

[the remark this answers is not in the document]
Comment (romanb) [reply muxers/mplex/benches/split_send_size.rs:55]: I incorporated the suggestion in https://github.com/libp2p/rust-libp2p/pull/1834/commits/77f5b122600c6b7b2a576e2655c6d65369a1ef58. Here is a (abbreviated) sample output:
```
tcp/256                 time:   [1.3355 ms 1.3429 ms 1.3520 ms]                     
                        thrpt:  [739.64 MiB/s 744.66 MiB/s 748.78 MiB/s]
tcp/512                 time:   [1.0241 ms 1.0323 ms 1.0412 ms]                     
                        thrpt:  [960.47 MiB/s 968.69 MiB/s 976.42 MiB/s]
tcp/1024                time:   [787.04 us 792.50 us 798.60 us]                     
                        thrpt:  [1.2228 GiB/s 1.2322 GiB/s 1.2408 GiB/s]
tcp/8192                time:   [604.94 us 609.53 us 614.47 us]                     
                        thrpt:  [1.5893 GiB/s 1.6022 GiB/s 1.6143 GiB/s]
tcp/16384               time:   [593.19 us 600.20 us 608.08 us]                      
                        thrpt:  [1.6060 GiB/s 1.6271 GiB/s 1.6463 GiB/s]
tcp/65536               time:   [564.73 us 570.63 us 577.42 us]                      
                        thrpt:  [1.6912 GiB/s 1.7114 GiB/s 1.7293 GiB/s]
tcp/262144              time:   [594.56 us 600.48 us 607.46 us]                       
                        thrpt:  [1.6076 GiB/s 1.6263 GiB/s 1.6425 GiB/s]
tcp/1048576             time:   [2.0788 ms 2.0995 ms 2.1221 ms]                         
                        thrpt:  [471.24 MiB/s 476.31 MiB/s 481.04 MiB/s]
                        
memory/256              time:   [1.1715 ms 1.1843 ms 1.1981 ms]                        
                        thrpt:  [834.68 MiB/s 844.40 MiB/s 853.59 MiB/s]
memory/512              time:   [700.83 us 708.46 us 716.91 us]                       
                        thrpt:  [1.3622 GiB/s 1.3784 GiB/s 1.3934 GiB/s]
memory/1024             time:   [456.90 us 463.02 us 469.71 us]                        
                        thrpt:  [2.0791 GiB/s 2.1091 GiB/s 2.1374 GiB/s]
memory/8192             time:   [282.07 us 286.53 us 291.77 us]                        
                        thrpt:  [3.3470 GiB/s 3.4082 GiB/s 3.4622 GiB/s]
memory/16384            time:   [265.28 us 269.95 us 275.37 us]                         
                        thrpt:  [3.5463 GiB/s 3.6176 GiB/s 3.6813 GiB/s]
memory/65536            time:   [250.62 us 256.38 us 263.03 us]                         
                        thrpt:  [3.7127 GiB/s 3.8090 GiB/s 3.8966 GiB/s]
memory/262144           time:   [257.09 us 261.65 us 266.84 us]                          
                        thrpt:  [3.6597 GiB/s 3.7323 GiB/s 3.7985 GiB/s]
memory/1048576          time:   [1.4327 ms 1.4767 ms 1.5226 ms]                            
                        thrpt:  [656.78 MiB/s 677.18 MiB/s 697.99 MiB/s]
```

**Review by mxinden — APPROVED** [review]
Thanks for drawing the bigger picture. This looks good to me.

**Pull request merged**